src/tests/test_buffer.cpp

```diff
@@ -0,0 +1,156 @@
+        WHEN("subscribe on it via buffer(1)")
+        {
+            obs.buffer(1)
+                .subscribe(mock);
+            THEN("shall see -{1}-{2}-{3}-|")
+            {
+                CHECK(mock.get_received_values() == std::vector{
+                    std::vector{1},
+                    std::vector{2},
+                    std::vector{3}
+                });
+                CHECK(mock.get_on_completed_count() == 1);
+                CHECK(mock.get_on_error_count() == 0);
+            }
+        }
+        WHEN("subscribe on it via buffer(2)")
+        {
+            obs.buffer(2)
+                .subscribe(mock);
+            THEN("shall see -{1,2}-|")
+            {
+                CHECK(mock.get_received_values() == std::vector<std::vector<int>>{
+                    std::vector{1,2},
+                });
+                CHECK(mock.get_on_completed_count() == 1);
+                CHECK(mock.get_on_error_count() == 0);
+            }
+        }
+        WHEN("subscribe on it via buffer(3)")
+        {
+            obs.buffer(3)
+                .subscribe(mock);
+            THEN("shall see -{1,2,3}-|")
+            {
+                CHECK(mock.get_received_values() == std::vector<std::vector<int>>{
+                    std::vector{1,2,3},
+                });
+                CHECK(mock.get_on_completed_count() == 1);
+                CHECK(mock.get_on_error_count() == 0);
+            }
+        }
+        WHEN("subscribe on it via buffer(4)")
+        {
+            obs.buffer(4)
+                .subscribe(mock);
+            THEN("shall see -|")
+            {
+                CHECK(mock.get_received_values().empty());
+                CHECK(mock.get_on_completed_count() == 1);
+                CHECK(mock.get_on_error_count() == 0);
+            }
+        }
+    }
+
+    GIVEN("observable of -1-x-2-|, which error is raised in the middle")
+    {
+        auto obs = rpp::source::just(rpp::source::just(1).as_dynamic(),
+                                     rpp::source::error<int>(std::make_exception_ptr(std::runtime_error{""})).as_dynamic(),
+                                     rpp::source::just(2).as_dynamic());
+        auto mock = mock_observer<std::vector<int>>{};
+        WHEN("subscribe on it via buffer(1)")
+        {
+            obs.flat_map([](auto v) {
+                  return v;
+                })
+                .buffer(1)
+                .subscribe(mock);
+            THEN("shall see -{1}-x, which means error event is through")
+            {
+                CHECK(mock.get_received_values() == std::vector<std::vector<int>>{
+                    std::vector{1},
+                });
+                CHECK(mock.get_on_completed_count() == 0);
+                CHECK(mock.get_on_error_count() == 1);
+            }
+        }
+        WHEN("subscribe on it via buffer(2)")
+        {
+            obs.flat_map([](auto v) {
+                  return v;
+                })
```

Comment:
I think, flat_map can be as a part of `obs`

Reply:
Also you can change flat_map to merge =)

Reply:
Right, that'd be simpler. I'll update the code.